File: src/main/java/com/vaadin/demo/component/sidenav/SideNavSuffix.java
package com.vaadin.demo.component.sidenav;

import com.vaadin.flow.component.badge.Badge;
import com.vaadin.flow.component.badge.BadgeVariant;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.icon.VaadinIcon;
import com.vaadin.flow.component.sidenav.SideNav;
import com.vaadin.flow.component.sidenav.SideNavItem;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.demo.component.sidenav.views.CalendarView;
import com.vaadin.demo.component.sidenav.views.InboxView;

@Route("side-nav-suffix")
public class SideNavSuffix extends Div {

    public SideNavSuffix() {
        // tag::snippet[]
        SideNav nav = new SideNav();

        SideNavItem inboxLink = new SideNavItem("Inbox", InboxView.class,
                VaadinIcon.ENVELOPE.create());
        Badge inboxCounter = new Badge("unread messages", 12);
        inboxCounter.addThemeVariants(BadgeVariant.FILLED,
                BadgeVariant.NUMBER_ONLY);
        inboxLink.setSuffixComponent(inboxCounter);

        SideNavItem calendarLink = new SideNavItem("Calendar",
                CalendarView.class, VaadinIcon.CALENDAR.create());
        Badge calendarNotification = new Badge("Upcoming appointment");
        calendarNotification.setIcon(VaadinIcon.BELL.create());
        calendarNotification.addThemeVariants(BadgeVariant.ERROR,
                BadgeVariant.ICON_ONLY);
        calendarLink.setSuffixComponent(calendarNotification);

        nav.addItem(inboxLink, calendarLink);
        // end::snippet[]

        Div navWrapper = new Div(nav);
        nav.setWidthFull();
        add(navWrapper);

        nav.getElement().executeJs("window.patchSideNavNavigation(this);"); // hidden-source-line

        this.addClassName("side-nav-sample");
    }

    public static class Exporter extends DemoExporter<SideNavSuffix> { // hidden-source-line
    } // hidden-source-line
}

File: frontend/demo/component/side-nav/side-nav-suffix.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/badge';
import '@vaadin/icon';
import '@vaadin/icons';
import '@vaadin/side-nav';
import { html, LitElement } from 'lit';
import { customElement } from 'lit/decorators.js';
import { patchSideNavNavigation } from 'Frontend/demo/component/side-nav/side-nav-helper'; // hidden-source-line
import { applyTheme } from 'Frontend/demo/theme';

@customElement('side-nav-suffix')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    applyTheme(root);
    return root;
  }

  /* prettier-ignore */ protected firstUpdated() { // hidden-source-line
    patchSideNavNavigation(this.shadowRoot!.querySelector('vaadin-side-nav')!); // hidden-source-line
  } // hidden-source-line

  protected override render() {
    return html`
      <div class="side-nav-sample">
        <div>
          <!-- tag::snippet[] -->
          <vaadin-side-nav style="width:100%">
            <vaadin-side-nav-item path="/inbox">
              <vaadin-icon icon="vaadin:envelope" slot="prefix"></vaadin-icon>
              Inbox
              <vaadin-badge number="12" theme="filled number-only" slot="suffix">
                unread messages
              </vaadin-badge>
            </vaadin-side-nav-item>
            <vaadin-side-nav-item path="/calendar">
              <vaadin-icon icon="vaadin:calendar" slot="prefix"></vaadin-icon>
              Calendar
              <vaadin-badge theme="error icon-only" slot="suffix">
                <vaadin-icon icon="vaadin:bell" slot="icon"></vaadin-icon>
                Upcoming appointment
              </vaadin-badge>
            </vaadin-side-nav-item>
          </vaadin-side-nav>
          <!-- end::snippet[] -->
        </div>
      </div>
    `;
  }
}
